### DISCUSSION TITLE
Need some hint to implement a "user finite element"

### DISCUSSION THREAD

**Discussion by Vincent-IUTRS-EOST** [idaholab/moose#20788] [Q&A General] [answered]
Dear MOOSE users,

I am currently working on a TTHM (thermo-hydro-mechanical) model with 2 temperatures, one for the solid grains of rocks and the other for the saturating fluid. My balance equations are implemented and OK, but I would like now to implement a specific finite element modeling the flow in a fracture, the permeability and the normal/tangential mechanical behavior of which depends on the fracture aperture. Here is a little figure for the sake of clarity:

![image](https://user-images.githubusercontent.com/103634204/163359828-622336e2-b2c2-4afa-90e7-ce32d19839e2.png)

As I understand the MOOSE philosophy, I cannot introduce a new HEX element for which I could declare dofs for each nodes, as I did before with another code called Code_Aster (https://www.code-aster.org).
 
I was thinking to split my mesh with "BreakMeshByBlockGenerator", with "add_transition_interface=true", but I am not sure (and in fact, don't see how) I will be able to calculate the normal displacement xi.n in my kernel calculating the 2D flow.

Does anyone have some hint or strategy to propose ?

**Comment by GiudGiud** [comment]
Hello

BreakMeshByBlockGenerator is for large scale subdivisions of the mesh. You are showing one element here, is this split a long fracture?
Without too much understanding of your problem, you will need to solve an equation for the displacement of the mesh. This solve can be done at the same time as the fluid flow solve. The 2D flow kernels can use the displacement in their definition. 
@WilkAndy any thoughts on this?

Guillaume

**Comment by Vincent-IUTRS-EOST** [comment, marked as answer]
Dear Guillaume,

Thank you for your answer. The element on the previous figure is a single finite element simulating the flow in a fracture which can be as wide as my mesh. I provide below another figure, trying to clarify the situation:

![image](https://user-images.githubusercontent.com/103634204/163471865-fb27218a-4aea-4e8f-b9ed-8f91a5ff44d6.png)

My problem is to calculate the displacement jump (xi+) - (xi-) (giving the fracture aperture and then the permeability of it) at the quadrature point of the 2D surface containing the 2D flow. 

A+

**Reply by GiudGiud** [reply]
Thanks for the additional explanation. 
@cpgr @RLiu2020 should also be knowledgable

**Reply by GiudGiud** [reply]
@roystgnr as well, this is kind of a libmesh question in a way

**Reply by roystgnr** [reply]
The libMesh answer, I'm afraid, is "TBD, but not available right now": https://github.com/libMesh/libmesh/discussions/3226

In the shorter term, though ... that mesh looks very nicely fracture-aligned.  Would it be sufficient to split it along the fracture, turning each hex containing the fracture into two hexes and a quad?  I'm afraid you'd need some additional data structure too (Our Elem::interior_map() could be used from each quad to access one of the neighboring hexes, but not both.  In hindsight that inadequacy of our boundary elements for internal boundaries has been a problem forever, not sure why I never noticed before.) but then you'd just evaluate the internal boundary condition (as well as the fracture flow) on a kernel on the quads.

**Comment by JerryLiu2023** [comment]
Sorry for late response. As Roy discussed, there are a lot development work. But at least, you can have the jump of displacement across the interface. Since you use  BreakMeshByBlockGenerator, your kernel is an interface kernel. In your interface kernel, use getmaterialproperty(displacement) (something likes this) to get displacement on the primary element side, and getneighbornaterialproperty(displacement) to get u for the corresponding other side. No matter you use displaced mesh or not, you Always get displacements. So, jump of u is understood. You already did this before using other codes, but just a reminder that If we don't apply pore pressure on these split two surfaces as appled surface pressure loadings(driving force for fracturing) from the solid equation of view), we may get zero gaps or even negative apertures. Thanks!

**Comment by Vincent-IUTRS-EOST** [comment]
Dear all,

Thanks a lot for your answers. As @roystgnr wrote, I have a tiny Python library that meshes my reservoir with HEX elements and apply a function to nodes coordinates afterwards, so that HEX elements are approximately aligned along the fracture. It is then rather easy to implement a small function that would split - as he suggests- by doing : HEX along fracture --> 2 x HEX+ 1 x QUAD. For the moment BreakMeshByBlockGenerator is only an option that I don't exclude.

Nevertheless, as @RLiu2020 noticed, one difficulty is to properly write the mechanical equilibrium of the fracture boundary since the pore pressure appears in the traction vector of the fracture boundaries. After some days of holiday, and to avoid entering deeply in the libmesh library, I was thinking to a new global strategy using the already-implemented features of MOOSE (or features that could be easily implemented in the MOOSE C++ "high" layer rather than the libmesh "down" layer which would cost me a lot of time). This strategy consists in creating a kind of "copy" of the fracture boundary variables, so that the entire fracture behavior (mechanical normal and tangential behaviors + 2D fluid flow) can be solved in the QUAD elements. I provide another scheme for the sake of clarity, and the global strategy written in a kind of MOOSE pseudocode: 

![image](https://user-images.githubusercontent.com/103634204/163961313-be6c81eb-2f55-4a19-8859-7dbb1517866b.png)

I would like to know, based on your experience and since I am using MOOSE only for 2 months (but with some numerical skills from the use of other FE codes before), if this approach may be interesting and may have a chance to work. I noticed that some difficulties may still arise since LinearNodalConstraint is - unfortunately - written for only one variable.... but I hope libmesh is friendly enough ;-) to provide me some interface like mesh.getDof(id,jvar) , giving me the (jvar)^th degree of freedom of node "id" , so that I can implement my own NodalConstraint object.

Vincent.

**Reply by roystgnr** [reply]
`mesh.node_ref(id).dof_number(system_number,variable_number,component_number)` will give you the global degree of freedom number for a particular component (some FE types have multiple) of a particular variable on a particular system.  Usually if you find yourself using that directly it means there's a better way to do things that you're missing.  I'm not sure what the better way might be at the Moose level here, though.
